pkg/common/vsphere/virtual_center.go

```diff
@@ -0,0 +1,289 @@
+	vc.connectMutex.Lock()
+	defer vc.connectMutex.Unlock()
+	log := vc.logger
+	// Set up the vc connection
+	err := vc.connect(ctx, false)
+	if err != nil {
+		log.Errorf("Cannot connect to vCenter with err: %v", err)
+		return nil, nil, err
+	}
+	return vc.vslmClient, vc.cnsClient, nil
+}
+
+// connect creates a connection to the virtual center host.
+func (vc *VirtualCenter) connect(ctx context.Context, requestNewSession bool) error {
+	log := vc.logger
+	// If client was never initialized, initialize one.
+	var err error
+	if vc.Client == nil {
+		vc.Client, err = vc.newClient(ctx)
+		if err != nil {
+			log.Errorf("failed to create govmomi client with err: %v", err)
+			return err
+		}
+	}
+	if vc.vslmClient == nil {
+		vslmVcClient, err := newVslmClient(ctx, vc.Client.Client, log)
+		if err != nil {
+			log.Errorf("failed to create VSLM client on vCenter host %q with err: %v", vc.Config.Host, err)
+			return err
+		}
+		vc.vslmClient = vslm.NewGlobalObjectManager(vslmVcClient)
+	}
+	if vc.cnsClient == nil {
+		vc.cnsClient, err = newCnsClient(ctx, vc.Client.Client, log)
+		if err != nil {
+			log.Errorf("failed to create CNS client on vCenter host %q with err: %v", vc.Config.Host, err)
+			return err
+		}
+	}
+	if !requestNewSession {
+		// If session hasn't expired, nothing to do.
+		sessionMgr := session.NewManager(vc.Client.Client)
+		// SessionMgr.UserSession(ctx) retrieves and returns the SessionManager's CurrentSession field
+		// Nil is returned if the session is not authenticated or timed out.
+		userSession, err := sessionMgr.UserSession(ctx)
+		if err != nil {
+			log.Errorf("failed to obtain user session with err: %v", err)
+			return err
+		}
+		// Got a valid session if userSession is non-nil
+		if userSession != nil {
+			return nil
+		}
+	}
+	// If session has expired, create a new instance.
+	log.Warnf("Creating a new client session as the existing session isn't valid or not authenticated")
+	vc.Client, err = vc.newClient(ctx)
+	if err != nil {
+		log.Errorf("failed to create govmomi client with err: %v", err)
+		return err
+	}
+
+	// Recreate vslmClient If created using timed out VC Client
+	vslmVcClient, err := newVslmClient(ctx, vc.Client.Client, vc.logger)
+	if err != nil {
+		log.Errorf("failed to create VSLM client with err: %v", err)
+		return err
+	}
+	vc.vslmClient = vslm.NewGlobalObjectManager(vslmVcClient)
+
+	// Recreate CNSClient If created using timed out VC Client
+	vc.cnsClient, err = newCnsClient(ctx, vc.Client.Client, vc.logger)
+	if err != nil {
+		log.Errorf("failed to create CNS client on vCenter host %v with err: %v", vc.Config.Host, err)
+		return err
+	}
+
+	return nil
+}
+
+// Disconnect disconnects the virtual center host connection if connected.
+func (vc *VirtualCenter) Disconnect(ctx context.Context) error {
+	vc.connectMutex.Lock()
+	defer vc.connectMutex.Unlock()
+	log := vc.logger
+	vc.disconnectVslm()
```

Comment:
connect creates the cns Client and vslm client inline.  Why is there a separate disconnect for each?  If you need it on disconnect, you should have it for connect or vice versa.